### PR TITLE
Feature/add build test

### PR DESCRIPTION
**Pull request by JulieSchramm**
## DESCRIPTION OF CHANGES: 
Add ``build.sh`` to test directory to provide the build step for automated testing.  A ``README.md`` file is included with information on how to run the build.sh script.

## TESTS CONDUCTED: 
build.sh script tested on Cheyenne (Intel and GNU), Hera (Intel) and Jet (Intel).  All builds succeeded.

## DEPENDENCIES:
None

## DOCUMENTATION:
Provided in the README.md file.  Will be included in the User's Guide when testing is added.

### CODE REVIEW

File: test/build.sh [hunk 1]
```diff
@@ -0,0 +1,146 @@
+#!/bin/bash -l
```
Comment (gsketefian) [inline test/build.sh]: Does it work without -l?  That would be better.

**Comment by gsketefian** [comment]
How is build.sh different from the devbuild.sh that @mkavulich put together?

File: test/build.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  exit 1
+}
+
+machines=( hera jet cheyenne )
+
+[[ $# -eq 0 ]] && usage
+if [ "$1" = "-h" ] ; then usage ; fi
+
+export machine=${1}
+machine=`echo "${machine}" | tr '[A-Z]' '[a-z]'`  # scripts in sorc need lower case machine name
```
Comment (christinaholtNOAA) [inline test/build.sh]: In general, the `$()` syntax is easier to read/debug.

File: test/build.sh [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+if [ "${machine}" == "cheyenne" ] ; then
+  compilers=( intel gnu )
+else
+  compilers=( intel )
+fi
+
+#-----------------------------------------------------------------------
+# Set some directories
+#-----------------------------------------------------------------------
+PID=$$
```
Comment (christinaholtNOAA) [inline test/build.sh:59]: Just for my enrichment, what does this syntax do?

File: test/build.sh [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                                 vcoord_gen )
+
+#-----------------------------------------------------------------------
+# Set up the build environment and run the build script.
+#-----------------------------------------------------------------------
+  for compiler in "${compilers[@]}"; do
+    BUILD_DIR=${TOP_DIR}/build_${compiler}
+    BUILD_OUTPUT=${BUILD_DIR}/build.out
+    BIN_DIR=${TOP_DIR}/bin_${compiler}
+    EXEC_DIR=${BIN_DIR}/bin
```
Comment (christinaholtNOAA) [inline test/build.sh:103]: Just curious why a subdirectory is needed.

**Comment by JulieSchramm** [comment]
The -l is handy if you call this script from crontab.  I can remove it, and
it can be added in the crontab file if someone wants to use it that way.

And how is the build.sh different from the devbuild.sh that @danrosen25 is
putting together?

File: test/build.sh [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    if [ $build_it -eq 0 ] ; then
+      . ${ENV_DIR}/build_${machine}_${compiler}.env
+#-----------------------------------------------------------------------
+# Remove old and bin directories if they exist
+#-----------------------------------------------------------------------
+      if [ -d "${BUILD_DIR}" ] ; then rm -rf ${BUILD_DIR} ; fi
+      if [ -d "${BIN_DIR}" ] ; then rm -rf ${BIN_DIR} ; fi
+      mkdir ${BUILD_DIR}
+      cd ${BUILD_DIR}
+      cmake .. -DCMAKE_INSTALL_PREFIX=../bin_${compiler}
```
Comment (christinaholtNOAA) [inline test/build.sh]: Can this line use `${BIN_DIR}` to reduce duplication of information?

**Review by christinaholtNOAA — APPROVED** [review]
Just a couple of very minor suggestions, and a quick question for my curiosity (hard to google).

**Comment by JeffBeck-NOAA** [comment]
@JulieSchramm, @danrosen25's PR only improves the devbuild.sh script that @mkavulich created.  But I think @gsketefian is asking why we need this script in addition to devbuild.sh.

File: test/build.sh (hunk 2 above)
Comment (JulieSchramm) [reply test/build.sh]: Agree, will change.

File: test/build.sh (hunk 3 above)
Comment (JulieSchramm) [reply test/build.sh:59]: ``$$`` is a process ID (PID) used to create a unique output filename each time the script is run so the previous output is saved.

File: test/build.sh (hunk 4 above)
Comment (JulieSchramm) [reply test/build.sh:103]: In the top-level CMakeLists.txt file, the install directory is set as ``bin`` under what  ``-DCMAKE_INSTALL_PREFIX`` is set to.  To make multiple builds and save all the executables for testing, I set ``-DCMAKE_INSTALL_PREFIX=../bin_${compiler}``, which creates ``../bin_intel/bin``, ``../bin_gnu/bin``, etc.

File: test/build.sh (hunk 3 above)
Comment (christinaholtNOAA) [reply test/build.sh:59]: Thanks, @JulieSchramm!

File: test/build.sh (hunk 5 above)
Comment (JulieSchramm) [reply test/build.sh]: Yes!

**Comment by JulieSchramm** [comment]
The build.sh script is for testing.  It builds all executables for all supported compilers on a given machine.  It checks for the existence of all executables and writes a pass/fail message at the end.  This will be the build step for Christina et al.'s testing.

**Comment by gsketefian** [comment]
I'm wondering if it's possible to call the `devbuild.sh` script within the `build.sh` script introduced in this PR so that the same build script that is normally used to build the code is used to test the build, i.e. `build.sh` would be a wrapper around `devbuild.sh`.  @danrosen25 is adding a bunch of new arguments to `devbuild.sh` in PR #155.  If those changes aren't enough, `devbuild.sh` can probably be tweaked further so that it can be called by `build.sh` (instead of `build.sh` issuing its own `cmake` and `make` commands).

File: test/build.sh (hunk 1 above)
Comment (danrosen25) [reply test/build.sh]: If possible I think it's best to design the script to be independent of the shell invocation selected. So it's best to eliminate any dependence on the environment settings at the time the script is called.

**Comment by JulieSchramm** [comment]
I would leave the decision about build.sh vs devbuild.sh to @christinaholtNOAA, as the group at GSL will be setting up the automated testing.

**Comment by christinaholtNOAA** [comment]
I have no strong opinions about one way or the other. If any users rely on `devbuild.sh` for building, I could see how calling it (instead of cmake directly) would essentially be testing an additional piece of used code, which is always better than not testing code that is used. Maybe that is enough to tip the scale toward calling `devbuild.sh` instead of using cmake?

**Comment by danrosen25** [comment]
Just a heads up the devbuild.sh changes will take arguments for
- `--platform=<PLATFORM>` , if omitted this will use the machine_setup.sh script
- `--compiler=<COMPILER>`, defaults to intel on cheyenne, hera, jet
- `--build-dir=<BUILD_DIR>`, the CMake build directory
- `--install-dir=<INSTALL_DIR>` the CMAKE_INSTALL_PREFIX directory, which is put into the build directory if you don't give a full path.

More can be added as needed. I'm trying to make a user|developer|sysadmin friendly build script. We needed a way to build different configurations of the model i.e. `--app=<APPLICATION>`

**Comment by JulieSchramm** [comment]
All builds passed on Cheyenne, Jet and Hera.

File: test/build.sh [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+if [ ! -f "$TEST_OUTPUT" ]; then
+   touch ${TEST_OUTPUT}
+fi
+
+cd ${TOP_DIR}
+
+ENV_DIR=${TOP_DIR}/env
+#-----------------------------------------------------------------------
+# Arrray of all executables built
+#-----------------------------------------------------------------------
```
Comment (JeffBeck-NOAA) [inline test/build.sh:77]: Typo: Arrray to Array.

File: test/build.sh [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    if [ -f ${exec_file} ]; then
+      echo "SUCCEED: ${compiler} executable file ${exec_file} exists" >> ${TEST_OUTPUT}
+    else
+      echo "FAIL: ${compiler} executable file ${exec_file} does NOT exist" >> ${TEST_OUTPUT}
+      let "n_fail=n_fail+1"
+    fi
+  done
+  #-----------------------------------------------------------------------
+  # Rename build and bin dirs for next build
+  #-----------------------------------------------------------------------
```
Comment (JeffBeck-NOAA) [inline test/build.sh]: I'm confused by this comment block above.  What code does it apply to?

File: test/build.sh (hunk 6 above)
Comment (JulieSchramm) [reply test/build.sh:77]: Good catch

File: test/build.sh (hunk 7 above)
Comment (JulieSchramm) [reply test/build.sh]: Removed

**Pull request merged**